### PR TITLE
Create exos

### PR DESCRIPTION
**Pull request by almerion**

### CODE REVIEW

File: src/ajax/send_exercice.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    foreach ($data as $key => $value) {
+        if ($canBeSend === true && isset($strJsonFileContents->{strtolower($value->{'type'})})) {
+            $properties = $value->{'properties'};
+            if (!(isset($properties->{'left'}) && isset($properties->{'top'}))) {
+                $left = getDoubleSize($properties->{'left'});
+                $top = getDoubleSize($properties->{'top'});
+                $maxTop = getDoubleSize($json->{'height'});
+                if (!($left > 0 && $left < 21 && $top > 0 && $top < $maxTop && $maxTop > 0 && $maxTop < 29.7)) {
+                    $canBeSend = false;
```
Comment (NaulaN) [inline src/ajax/send_exercice.php:44]:
```suggestion
                    return false;
```

File: src/ajax/send_exercice.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+function canBeSend($json) {
+    $canBeSend = true;
+    $data = $json->{'elements'};
+
+    $strJsonFileContents = json_decode(file_get_contents("../../res/exerciceOptions.json"));
+
+    foreach ($data as $key => $value) {
+        if ($canBeSend === true && isset($strJsonFileContents->{strtolower($value->{'type'})})) {
```
Comment (NaulaN) [inline src/ajax/send_exercice.php:37]: useless variable
```suggestion
        if (isset($strJsonFileContents->{strtolower($value->{'type'})})) {
```

File: src/ajax/send_exercice.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                $left = getDoubleSize($properties->{'left'});
+                $top = getDoubleSize($properties->{'top'});
+                $maxTop = getDoubleSize($json->{'height'});
+                if (!($left > 0 && $left < 21 && $top > 0 && $top < $maxTop && $maxTop > 0 && $maxTop < 29.7)) {
+                    $canBeSend = false;
+                }
+            }
+        }
+    }
+    return $canBeSend;
```
Comment (NaulaN) [inline src/ajax/send_exercice.php:49]:
```suggestion
    return true;
```

File: src/css/exercices-page.css [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    background-color: white;
+    display: flex;
+    justify-content: center;
+}
+#preview {
+    position: relative;
+    height: 5cm;
+    width: 20cm;
+    border: dashed red 2px;
+    background-color: white;
```
Comment (NaulaN) [inline src/css/exercices-page.css:53]: Need to move on `white.css` theme

File: src/css/exercices-page.css [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+.img-input {
+    width: 20px;
+    height: 20px;
+    border: 2px solid #1c1b1b;
```
Comment (NaulaN) [inline src/css/exercices-page.css:58]: All about colors need to move on `white.css` or `dark.css`

File: src/css/exercices-page.css [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+}
+.img-input::before {
+    content: '';
+    position: absolute;
+    top: 25%;
+    left: 25%;
+    width: 40%;
+    height: 40%;
+    border: 1px dashed #1c1b1b;
```
Comment (NaulaN) [inline src/css/exercices-page.css:67]: Same comment

File: src/css/themes/white.css [hunk 7]
```diff
@@ -38,6 +38,10 @@
     background-color: var(--white3);
 }
 
+#header-exercises {
+    background-color: var(--white5);
+}
```
Comment (NaulaN) [inline src/css/themes/white.css:43]: Just that ?
Sad moment

File: src/html/exercices.html [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE html>
+<html lang="fr" id="html-exercises">
+<head>
+    <meta charset="UTF-8">
+    <title>exercices</title>
+    <link rel="stylesheet" href="../css/global.css">
+    <link rel="stylesheet" href="../css/exercices-page.css">
```
Comment (NaulaN) [inline src/html/exercices.html:7]: Need to add `<link rel="stylesheet" href="../css/white.css">`

File: src/index.php [hunk 9]
```diff
@@ -1,28 +1,33 @@
 <?php
     ini_set('display_errors', 1);
 
-    require_once "./modules/mod_user/mod_user.php";
-    require_once "./modules/mod_home/mod_home.php";
-    require_once "./connexion.php";
+        require_once "./modules/mod_user/mod_user.php";
+        require_once "./modules/mod_home/mod_home.php";
+        require_once "./connexion.php";
```
Comment (NaulaN) [inline src/index.php:6]: Why you put tabulation here ?

File: src/js/css_loader.js [hunk 10]
```diff
@@ -6,6 +6,7 @@ const pathsCSS =
     {
         "loginRegister": "./css/register-login.css",
         "home": "./css/home-page.css",
+        "exercices" : "./css/exercices-page.css"
```
Comment (NaulaN) [inline src/js/css_loader.js:9]: Useless things right now with pop in method

File: src/js/css_loader.js [hunk 11]
```diff
@@ -29,6 +30,8 @@ window.onload = function () {
                 createLinkCSS(pathsCSS['loginRegister']);
             } else if (valueVar === 'home') {
                 createLinkCSS(pathsCSS["home"]);
+            } else if (valueVar === "exercices") {
+                createLinkCSS(pathsCSS["exercices"]);
```
Comment (NaulaN) [inline src/js/css_loader.js:34]: Same here

File: src/js/exercices.js [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    idCategorie: "1"
+};
+
+
+title.addEventListener('input', (ev) => {
+    page.title = title.value;
+    jsonOutput.setAttribute('value', JSON.stringify(page));
+})
+document.getElementsByTagName('body')[0].addEventListener('keydown' , (ev) => {
+    if (ev.key === "Backspace" && selectedItem != null) {
```
Comment (NaulaN) [inline src/js/exercices.js:29]:
```suggestion
    if (ev.key === "Delete" && selectedItem != null) {
```
It's better with `supp` key

File: src/js/exercices.js [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        let id = selectedItem.getAttribute('value');
+        page.elements = page.elements.filter((el) => {
+            return id != el.id;
+        })
```
Comment (NaulaN) [inline src/js/exercices.js:33]:
```suggestion
        page.elements = page.elements.filter((el) => id != el.id);
```

File: src/js/exercices.js [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        mooveX = ev.offsetX;
+        mooveY = ev.offsetY;
+        draggedElement = ev.target;
+        draggedElement.classList.add('preview-elements-moving');
+    }
+})
+preview.addEventListener("mousemove", (ev) => {
+    if (draggedElement != null) {
+        let elementHeight = parseFloat(window.getComputedStyle(draggedElement, null).getPropertyValue('height').split('px')[0]);
+        let elementWidth = parseFloat(window.getComputedStyle(draggedElement, null).getPropertyValue('width').split('px')[0]);
```
Comment (NaulaN) [inline src/js/exercices.js:50]:
```suggestion
        let elementHeight = parseFloat(window.getComputedStyle(draggedElement, null).getPropertyValue('height').replace('px', ''));
        let elementWidth = parseFloat(window.getComputedStyle(draggedElement, null).getPropertyValue('width').replace('px', ''));
```

File: src/js/exercices.js [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        const bound = preview.getBoundingClientRect();
+        const relativeMousePositions = getRelativePositionsMouse(bound, ev);
+        let mousePosRelativelyToPreviewX = relativeMousePositions.posX - mooveX;
+        let mousePosRelativelyToPreviewY = relativeMousePositions.posY - mooveY;
+
+        if ((mousePosRelativelyToPreviewY > 0 && mousePosRelativelyToPreviewX > 0)
+            && (mousePosRelativelyToPreviewY + elementHeight < bound.height && mousePosRelativelyToPreviewX + elementWidth < bound.width)) {
+            draggedElement.style.top = 0.0264583333 * mousePosRelativelyToPreviewY + 'cm';
+            draggedElement.style.left = 0.0264583333 *mousePosRelativelyToPreviewX + 'cm';
```
Comment (NaulaN) [inline src/js/exercices.js:59]: `0.0264583333` i don't understand whats correspond this value ?
On my opinion, this value must be put into a variable for be more readable

File: src/js/exercices.js [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+    const img = document.createElement("div");
+    img.classList.add('img-input');
+    return img;
+}
+
+
+function clearOptionAside() {
+    for (let i = 0; i < optionAside.children.length; i++) {
+        optionAside.removeChild(optionAside.children[i]);
+    }
```
Comment (NaulaN) [inline src/js/exercices.js:114]:
```suggestion
    for (let child of optionAside.children) {
        optionAside.removeChild(child);
    }
```

File: src/js/exercices.js [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+            let options = createOptions(parameters[keys[i]], element, keys[i]);
+            optionAside.appendChild(options);
+        }
+    }
+}
+
+function getOptions(tagName) {
+    let key = Object.keys(rawAllOptions).find(el => {
+        return el.toLowerCase() === tagName.toLowerCase();
+    })
```
Comment (NaulaN) [inline src/js/exercices.js:131]:
```suggestion
    let key = Object.keys(rawAllOptions).find(el => el.toLowerCase() === tagName.toLowerCase());
```

File: src/js/exercices.js [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+        "posY": (event.clientY - bound.top)
+    };
+}
+
+function updateObject(element) {
+    let elementId = element.getAttribute('value');
+    if (typeof elementId === "string") {
+        let objectToUpdate = page.elements.find((el) => {
+            return el.id === elementId;
+        })
```
Comment (NaulaN) [inline src/js/exercices.js:182]:
```suggestion
        let objectToUpdate = page.elements.find((el) => el.id === elementId);
```

**Review by NaulaN — CHANGES_REQUESTED** [review]

File: src/index.php (hunk 9 above)
Comment (almerion) [reply src/index.php:6]: because [https://www.youtube.com/watch?v=C2iK35Mtgbk](url)

Comment (NaulaN) [reply src/index.php:6]: 🚀🚀🚀🚀🚀🚀🚀

File: src/js/css_loader.js (hunk 10 above)
Comment (almerion) [reply src/js/css_loader.js]: <object width="560" height="315" src="https://www.youtube.com/embed/LdzH8wPuQfI"></object>

Comment (almerion) [reply src/js/css_loader.js]: done

**Comment by NaulaN** [comment]
You need to add `<link rel="...." href="...." id="theme">` and edit `toggle-theme.js`

**Pull request merged**